Add service block to formula api

apainintheneck opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This fixes #14925.

The idea here is to include the service block info in the formula API JSON so that it can be loaded from the API. This will allow us to more consistently build service files and will mean that `brew services` doesn't have to be updated to work with the new API.

The first thing that was added was the ability to serialize and deserialize service blocks. We needed the ability to replace all local paths with placeholders when building the API JSON and then replace those placeholders with local paths when loading from the API.

The second thing was adding this service block JSON to the API and loading it from the API.

The other commits are just tests.

apainintheneck commented on Library/Homebrew/formula.rb:
```diff
@@ -1043,7 +1043,7 @@ def systemd_timer_path
     return unless service?
 
-    Homebrew::Service.new(self, &self.class.service)
+    @service ||= Homebrew::Service.new(self, &self.class.service)
```
Was there any special reason why this wasn't memoized before?

MikeMcQuaid replied:
Probably not.

apainintheneck commented on Library/Homebrew/formulary.rb:
```diff
@@ -263,6 +263,21 @@ def install
+        service_hash = Homebrew::Service.deserialize(service_hash)
+        run_params = service_hash.delete("run")
+        service do
+          if run_params.is_a?(Hash)
+            run(**run_params)
+          else
+            run run_params
+          end
```
`Service#run` accepts either one param or kwargs so needs to be handled separately.

SMillerDev replied:
How about keep_alive?

apainintheneck replied:
https://github.com/Homebrew/brew/blob/8f6276ed5f1c7b8293128a481b175611cb2980c8/Library/Homebrew/service.rb#L121-L141

That accepts a string or a hash but always gets stored internally as a hash. It's not a problem if you just pass that hash to it then. It'd be trickier if it accepted kwargs.

apainintheneck commented on Library/Homebrew/service.rb:
```diff
@@ -45,6 +45,9 @@ def f
       ).returns(T.nilable(Array))
     }
     def run(command = nil, macos: nil, linux: nil)
+      # Save parameters for serialization
+      @run_params ||= command || { macos: macos, linux: linux }.compact
+
```
Open to other ideas of how to handle this. Couldn't think of anything better.

MikeMcQuaid replied:
Seems fine to me!

apainintheneck commented on Library/Homebrew/service.rb:
```diff
@@ -320,7 +323,7 @@ def parse_cron(cron_statement)
       parsed
     end
 
-    sig { params(variables: T::Hash[String, String]).returns(T.nilable(T::Hash[String, String])) }
+    sig { params(variables: T::Hash[Symbol, String]).returns(T.nilable(T::Hash[Symbol, String])) }
```
This is only used with `Symbols` in `homebrew/core`.

```
/u/l/H/L/T/h/homebrew-core (master|✔) $ git grep environment_variables
Formula/apache-archiva.rb:    environment_variables ARCHIVA_BASE: var/"archiva"
Formula/appium.rb:    environment_variables PATH: std_service_path_env
Formula/clipper.rb:    environment_variables LANG: "en_US.UTF-8"
Formula/couchdb-lucene.rb:    environment_variables HOME: "~"
Formula/davmail.rb:    environment_variables PATH: std_service_path_env
Formula/docker-machine.rb:    environment_variables PATH: std_service_path_env
Formula/docuum.rb:    environment_variables PATH: std_service_path_env
Formula/dovecot.rb:    environment_variables PATH: std_service_path_env
Formula/ejabberd.rb:    environment_variables HOME: var/"lib/ejabberd"
Formula/etcd.rb:    environment_variables ETCD_UNSUPPORTED_ARCH: "arm64" if Hardware::CPU.arm?
Formula/gitlab-runner.rb:    environment_variables PATH: std_service_path_env
Formula/hbase.rb:    environment_variables HBASE_HOME:              opt_libexec,
Formula/httpd.rb:    environment_variables PATH: std_service_path_env
Formula/influxdb.rb:    environment_variables INFLUXD_CONFIG_PATH: etc/"influxdb2/config.yml"
Formula/isync.rb:    environment_variables PATH: std_service_path_env
Formula/jack.rb:    environment_variables PATH: "/usr/bin:/bin:/usr/sbin:/sbin:#{HOMEBREW_PREFIX}/bin"
Formula/kettle.rb:    environment_variables KETTLE_HOME: etc/"kettle"
Formula/launch_socket_server.rb:    environment_variables LAUNCH_PROGRAM_TCP_ADDRESS: "127.0.0.1:8080"
Formula/libvirt.rb:    environment_variables PATH: HOMEBREW_PREFIX/"bin"
Formula/nzbget.rb:    environment_variables PATH: "#{HOMEBREW_PREFIX}/bin:/usr/bin:/bin:/usr/sbin:/sbin"
Formula/offlineimap.rb:    environment_variables PATH: std_service_path_env
Formula/payara.rb:    environment_variables GLASSFISH_HOME: opt_libexec/"glassfish"
Formula/podman.rb:    environment_variables PATH: std_service_path_env
Formula/postgresql@15.rb:    environment_variables LC_ALL: "C"
Formula/rabbitmq.rb:    environment_variables PATH:          "#{HOMEBREW_PREFIX}/sbin:/usr/sbin:/usr/bin:/bin:#{HOMEBREW_PREFIX}/bin",
Formula/runit.rb:    environment_variables PATH: "/usr/bin:/bin:/usr/sbin:/sbin:#{opt_bin}"
Formula/wildfly-as.rb:    environment_variables JBOSS_HOME: opt_libexec, WILDFLY_HOME: opt_libexec
Formula/znapzend.rb:    environment_variables PATH: std_service_path_env
Formula/zookeeper.rb:    environment_variables SERVER_JVMFLAGS: "-Dapple.awt.UIElement=true"
Formula/zrepl.rb:    environment_variables PATH: std_service_path_env
```

apainintheneck commented on Library/Homebrew/service.rb:
```diff
@@ -497,5 +500,87 @@ def to_systemd_timer
+      end
+    end
+
+    # Recursively substitute path and string values in a nested JSON-like data structure.
```
Could use a better comment.

apainintheneck commented on Library/Homebrew/service.rb:
```diff
@@ -497,5 +500,87 @@ def to_systemd_timer
+      @eval_service_block = true
+    end
+
+    # Recursively substitute path and string values in a nested JSON-like data structure.
```
Could use a better comment.

apainintheneck commented:
Here's my testing script for checking that the service files were being generated correctly and that the serialization was consistent.

- Script: https://gist.github.com/apainintheneck/6ed138ac7214788cd20bb7da82f7aa36

MikeMcQuaid commented on Library/Homebrew/service.rb:
```diff
@@ -497,5 +500,93 @@ def to_systemd_timer
+    # - Replace each String prefixed with ":" with a Symbol
+    # - Replace "$HOMEBREW_PREFIX" with the local HOMEBREW_PREFIX environment variable
+    # - Replace "$HOME" with the local home directory
+    def self.deserialize(elem)
```
`from_hash` feels more typical naming here

apainintheneck replied:
I guess I'd expect `from_hash` to create a service object which is not what's happening here. I'm not really happy with the naming here either though.

MikeMcQuaid commented on Library/Homebrew/service.rb:
```diff
@@ -497,5 +500,93 @@ def to_systemd_timer
+    # - Replace each Symbol with a String prefixed by ":"
+    # - Replace the local HOMEBREW_PREFIX with "$HOMEBREW_PREFIX"
+    # - Replace the local home directory with "$HOME"
+    def serialize(elem = to_h)
```
Can this get a type signature?

apainintheneck replied:
We could always slap on something generic but I'm not sure how useful a specific type signature would be here. The only other option would be refactoring this into smaller functions.

@dduugg How would you approach adding typing to a recursive function like this?

dduugg replied:
If I'm reading the method documentation correctly, the best you can probably do is:
```ruby
sig { params(elem: T.untyped).returns(T::Hash[String, T.untyped]) }
```
Trying to type an undefined shape generally isn't worth the effort. `JSON.dump` is probably analogous here, and sorbet just leaves it completely [untyped](https://github.com/sorbet/sorbet/blob/60fb704/rbi/stdlib/json.rbi#L130-L138).

dduugg replied:
Oh, another pattern I've seen involves typing the input, something like:
```ruby
sig { params(elem: T::Hash[Symbol, T.untyped]).returns(T::Hash[String, T.untyped]) }
```
and then have `serialize` call a private `_serialize` method that handles the recursive calls. (Ofc, if you know more about the shape of the input/output you could flesh out the type annotations accordingly.)

apainintheneck replied:
Addressed in https://github.com/Homebrew/brew/pull/15007/commits/bc85857e1341eef2e105cbacd3e5f1e4eaaf47a0

MikeMcQuaid commented on Library/Homebrew/service.rb:
```diff
@@ -497,5 +500,93 @@ def to_systemd_timer
+    # - Replace each String prefixed with ":" with a Symbol
+    # - Replace "$HOMEBREW_PREFIX" with the local HOMEBREW_PREFIX environment variable
+    # - Replace "$HOME" with the local home directory
+    def self.deserialize(elem)
```
Can this get a type signature?

MikeMcQuaid commented on Library/Homebrew/service.rb:
```diff
@@ -497,5 +500,93 @@ def to_systemd_timer
+    def serialize(elem = to_h)
+      case elem
+      when String, Pathname
+        elem.to_s.gsub(HOMEBREW_PREFIX, HOMEBREW_PREFIX_PLACEHOLDER)
+            .gsub(Dir.home, "$HOME")
```
Feels like having a shared method to do this in `Utils` or something that can be used elsewhere will help us avoid forgetting to add more of these.

apainintheneck replied:
How would that method work? The way I see it we don't always apply or need to apply the same substitutions everywhere. We do have the `keg_relocate` code but something like that seems heavy handed for this specific case.

CC: @Rylan12 since you mentioned something similar in another thread

apainintheneck replied:
We could also do the substitutions right before or immediately after writing the `cask.json` or `formula.json` file if we wanted to have something super generic and only do the substitutions in one place.

Rylan12 replied:
I think one option is to have a method that just does all of the replacements we deem necessary (essentially `Cask::Cask.to_h_string_gsubs` before it was removed):

```ruby
def replace_strings(string)
  string.gsub(HOMEBREW_PREFIX, HOMEBREW_PREFIX_PLACEHOLDER)
        .gsub(HOME, HOME_PLACEHOLDER)
```

And then we can either selectively call this where it's needed or just recursively call this on the hash generated by `Formula#to_hash` before returning

MikeMcQuaid replied:
> just recursively call this on the hash generated by `Formula#to_hash` before returning

Something like this makes most sense to me.

I don't think it makes sense to scatter the replacements all over the codebase when they are only needed when reading/writing hashes.

CC @issyl0 as this affects https://github.com/Homebrew/brew/pull/15004 and possibly other PRs you have open

apainintheneck replied:
> Yeah, the only reason I can see for not doing that is that there is at least one place we decided we didn't want it to be replaced in casks (#14595)

Yep, this is the exact reason I'm worried about applying it more generally. Also, if we know the substitutions are not needed for most of the formula hash we're just doing extra work for no reason.

MikeMcQuaid replied:
Yeh, I forgot this method was changed in https://github.com/Homebrew/brew/pull/14615 to use monkeypatches to avoid the need for as many explicit substitutions.

I think it'd be nicer to use this approach instead wherever possible.

apainintheneck replied:
That could work for the hash generation but we'd still need to substitute things afterwards when deserializing. I think that the generating hash code for casks can probably just be converted into a generic mixin.

apainintheneck replied:
Addressed in https://github.com/Homebrew/brew/pull/15007/commits/38146893c3966d0ad1a359f866df4dbf49b9e176

MikeMcQuaid reviewed (approved):
Looks good! Some style thoughts.

apainintheneck commented:
Thanks for the feedback. I think that I'll rewrite the recursive (de)serialization methods to get rid of the recursion. It'll allow me to add better type annotations and I don't think the special handling around symbols will be necessary anymore which will be another bonus. It will be a little more verbose but that's probably fine.

Bo98 commented:
Ping me when this is near final and I'll review this.

apainintheneck commented:
I changed the serialization methods to not be recursive to make it easier to enable typing. I also changed the paths substitutions when serializing to use the same method as we already use with casks. This involves monkey patching the `$HOME` and `$HOMEBREW_PREFIX` values when generating the hash. Beyond that I updated the script from before.

- [Script](https://gist.github.com/apainintheneck/6ed138ac7214788cd20bb7da82f7aa36)

The generated service files are still the same but the serialization is different. That is because I'm not converting the `Pathname` objects to `String`s anymore because that happens naturally when converting it to JSON later on.

You can see the different serialization results [here](https://gist.github.com/apainintheneck/8b441c2a63d41dd88f68ddd5d41e36fb).

I still need to do more testing around the `generate-*-api` commands though to make sure they're still working correctly. Also, will there be any problems with monkey patching those common paths when generating the formula api JSON? I can't think of anything but I might have missed something.

apainintheneck commented:
@Bo98 This is ready for final review.

apainintheneck commented on Library/Homebrew/test/cask/cask_spec.rb:
```diff
@@ -337,6 +337,8 @@
       expect(JSON.pretty_generate(h["variations"])).to eq expected_sha256_variations.strip
     end
 
+    # @note The calls to `APIHashable.generating_hash!` and `APIHashable.generated_hash!`
+    #   are not idempotent so they can only be used in one test.
```
This comment should get updated. It's more that you can't have multiple tests for the same class.

apainintheneck replied:
Maybe I should add it to the `APIHashable` file instead?

apainintheneck replied:
Sorry, this was more a note to self. 😁

apainintheneck replied:
Addressed in https://github.com/Homebrew/brew/pull/15007/commits/801ee5e4744ba4f046e01d442ffd749af2d163a5

MikeMcQuaid commented on Library/Homebrew/service.rb:
```diff
@@ -45,6 +45,9 @@ def f
     }
     def run(command = nil, macos: nil, linux: nil)
+      # Save parameters for serialization
+      @run_params ||= command || { macos: macos, linux: linux }.compact
```
```suggestion
      @run_params ||= command 
      @run_params ||= { macos: macos, linux: linux }.compact
```

feels a little bit more typical/idiomatic/easy to follow

apainintheneck replied:
Addressed in https://github.com/Homebrew/brew/pull/15007/commits/801ee5e4744ba4f046e01d442ffd749af2d163a5

apainintheneck commented on Library/Homebrew/service.rb:
```diff
@@ -497,5 +500,101 @@ def to_systemd_timer
+      end
+
+      %w[interval cron launch_only_once require_root restart_delay macos_legacy_timers sockets].each do |key|
+        next if (value = api_hash[key]).nil?
```
It's more that `false` can be a valid value here so `nil` is the only value we really know is invalid.

MikeMcQuaid replied:
All good!

apainintheneck commented on Library/Homebrew/cask/cask.rb:
```diff
@@ -16,11 +17,9 @@ class Cask
-    HOMEBREW_PREFIX_PLACEHOLDER = "$HOMEBREW_PREFIX"
     APPDIR_PLACEHOLDER = "$APPDIR"
 
     # TODO: can be removed when API JSON is regenerated with HOMEBREW_PREFIX_PLACEHOLDER.
```
That was updated in another PR. Should be fine once this is merged into master.

apainintheneck replied:
Addressed in https://github.com/Homebrew/brew/pull/15007/commits/6ccf21c2c9a563e95476d31d12c1c04eedc3deed

apainintheneck commented on Library/Homebrew/cask/cask_loader.rb:
```diff
@@ -329,8 +329,8 @@ def from_h_string_gsubs(string, appdir)
-              .gsub(Cask::HOMEBREW_PREFIX_PLACEHOLDER, HOMEBREW_PREFIX)
+              .gsub(HOMEBREW_HOME_PLACEHOLDER, Dir.home)
+              .gsub(HOMEBREW_PREFIX_PLACEHOLDER, HOMEBREW_PREFIX)
               .gsub(Cask::APPDIR_PLACEHOLDER, appdir)
```
Doesn't matter to me. Do you feel strongly either way?

apainintheneck replied:
Addressed in https://github.com/Homebrew/brew/pull/15007/commits/801ee5e4744ba4f046e01d442ffd749af2d163a5

apainintheneck commented on Library/Homebrew/service.rb:
```diff
@@ -497,5 +500,101 @@ def to_systemd_timer
+    sig { params(api_hash: Hash).returns(Hash) }
+    def self.deserialize(api_hash)
+      hash = {}
+      hash[:run] =
+        case api_hash["run"]
+        when Hash
+          api_hash["run"].to_h do |key, array|
+            [
+              key.to_sym,
+              array.map { |value| replace_placeholders(value) },
+            ]
+          end
+        when Array
+          api_hash["run"].map { |value| replace_placeholders(value) }
+        end
```
I prefer to have the whole case statement at the same indentation but the use of the method method is a good point.

apainintheneck replied:
Addressed in https://github.com/Homebrew/brew/pull/15007/commits/801ee5e4744ba4f046e01d442ffd749af2d163a5

apainintheneck commented on Library/Homebrew/service.rb:
```diff
@@ -497,5 +500,101 @@ def to_systemd_timer
+      eval_service_block
+
+      cron_string = if @cron.present?
+        [:Minute, :Hour, :Day, :Month, :Weekday]
+          .map { |key| @cron[key].to_s }
+          .join(" ")
```
I prefer it like it is.

apainintheneck commented on Library/Homebrew/cask/cask_loader.rb:
```diff
@@ -329,8 +329,8 @@ def from_h_string_gsubs(string, appdir)
+              .gsub(HOMEBREW_HOME_PLACEHOLDER, Dir.home)
+              .gsub(HOMEBREW_PREFIX_PLACEHOLDER, HOMEBREW_PREFIX)
               .gsub(Cask::APPDIR_PLACEHOLDER, appdir)
               .gsub(Cask::HOMEBREW_OLD_PREFIX_PLACEHOLDER, HOMEBREW_PREFIX)
```
Addressed in https://github.com/Homebrew/brew/pull/15007/commits/6ccf21c2c9a563e95476d31d12c1c04eedc3deed

apainintheneck commented:
I'm now happy with this as it is right now. I'll wait another day for any final reviews and then merge it in sometime tomorrow.

As a sanity check I created another script that's a sort of `brew readable` but for the generated API. Might be useful for other PRs too.
- [Readall Script](https://gist.github.com/apainintheneck/ce23cf2ca96f0d5ff988f9ab16a7f4aa)

MikeMcQuaid reviewed (approved):
Looks great, thanks again @apainintheneck!

MikeMcQuaid commented:
> As a sanity check I created another script that's a sort of `brew readall` but for the generated API. Might be useful for other PRs too.

@apainintheneck 💯% yes! Would be nice to get this as a CI check in Homebrew/brew (not in/blocking this PR)? How long does it take to run for you?
